Controller support for broadcaster BIG commands and BIS events

Thalley opened the pull request:
Adds support for the BIG commands as broadcaster and enqueuing the empty BIS PDU.

Thalley committed:
Bluetooth: controller: Implements ULL handling for BIG create command

Implements handling of the BIG create command in the upper link layer.
Does not yet handle sending any BIS events.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: controller: BIG create sends empty BIS PDUs (ULL)

When a BIG is create, it will now start sending empty BIS PDUs;
at least from an ULL perspective, as the LLL support is still
missing.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: controller: Implements BIG terminate command

Implements the BIG terminate command that terminates a
BIG and returns the appropriate events.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
This PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/29785 as well as https://github.com/zephyrproject-rtos/zephyr/pull/29448 before it can compile without warnings.

Thalley committed:
Bluetooth: controller: Add initial basic ISO broadcast test

Adds a initial ISO broadcast test case that simply
creates, waits and terminates a BIG using the ll API.

The newly created test directory can be expanded to contain
connected ISO (CIS) tests as well in the future.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Buildkite fails on a non-related regression as far as I can tell. To be examined.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -10,28 +10,396 @@
+static struct ll_adv_iso ll_adv_iso[CONFIG_BT_CTLR_ADV_SET];
+static void *adv_iso_free;
+
+static MFIFO_DEFINE(iso_tx, sizeof(struct lll_tx),
+		    CONFIG_BT_CTLR_ISO_TX_BUFFERS);
+
+/* TODO: Share between BIS and CIS */
+static struct {
+	void *free;
+	uint8_t pool[CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE *
+			CONFIG_BT_CTLR_ISO_TX_BUFFERS];
+} mem_iso_tx;
```
This should really in a file that is common for BIS and CIS (cf. #30000, which creates such a file).

wopu-ot replied:
#30000 is merged now.

Thalley replied:
Added a commit in this PR that moved it to `ull_iso.c`

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -10,28 +10,396 @@
+	return 0;
+}
+
+int ull_adv_iso_reset(void)
```
I think a call to this function is missing in ull.c

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -10,28 +10,396 @@
+			return BT_HCI_ERR_INVALID_PARAM;
+		}
+
+		if (phy > 3) {
```
phy bit 2 would be coded PHY, so values > 3 would be legal. Consider using defines here (cf. defines `BT_HCI_LE_EXT_SCAN_PHY_...`)

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
I guess this way of allocating is okay for this very first iteration. However, I think we should discuss how ISO buffers should be handled in the final BIS/CIS implementation to make sure we do not go in a direction now that will make it difficult to do things right in the end.

Thalley replied:
I agree. This way is heavily inspired of how it's done for connection PDUs. Synchronizing how it's done for connection and ISO packets seems alright to me.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -10,28 +10,396 @@
+	if (adv_iso_empty) {
+		memset(adv_iso_empty, 0, sizeof(*adv_iso_empty));
+		adv_iso_empty->hci_handle = hci_handle;
+		*handle = idx;
```
`idx` is always incremented to `CONFIG_BT_CTLR_ADV_SET`, so I don't think this does what it should do.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_types.h:
```diff
@@ -60,5 +61,40 @@ struct ll_adv_sync_set {
+	uint8_t  bcode[16];
+
+	struct node_rx_hdr node_rx_complete;
+	struct {
+		struct node_rx_hdr node_rx_hdr_terminate;
+		uint8_t reason;
+	} node_rx_terminate;
```
Please add alignment and comments here similar to `node_rx_lost` in `ll_sync_set`.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -62,11 +430,62 @@ uint8_t ll_big_test_create(uint8_t big_handle, uint8_t adv_handle,
+			  TICKER_ID_ADV_ISO_BASE + adv_iso->bis_handle,
+			  ticker_op_stop_cb, adv_iso);
+
+	adv_iso->node_rx_terminate.reason = reason;
```
I think reason could be assigned directly to `node_rx->pdu` below instead of going through `node_rx_terminate_reason`. The `reason` field should alias `node_rx->pdu` anyway.

Thalley replied:
Sure. Personally I think assigning values to unnamed values (e.g. `node_rx->pdu`) is less readable than assigning the value to e.g. `adv_iso->node_rx_terminate.reason`, but it will be more consistent to other uses.

wopu-ot reviewed (changes requested):
Nice work, but with a PR this size comments are almost inevitable :-)

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_iso/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+# SPDX-License-Identifier: Apache-2.0
```
I was actually thinking that we'd reuse the same directory with several separated tests like what is done in the `bsim_test_app` directory. We can split it no problem, if that is still wanted.

wopu-ot replied:
On the one hand, I think splitting would be good to keep the size of the tests down. On the other hand having the same directory makes sharing of code between test cases easier. I am okay with keeping it as it is now, we can always move later.

wopu-ot commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -193,6 +193,23 @@ config BT_CTLR_TX_BUFFER_SIZE
+	int "Number of Isochronous Tx buffers"
+	default 7 if BT_HCI_RAW
+	default 3
+	range 1 19
```
How many buffers are needed depends very much on the use case. I would avoid putting a limit unless something necessarily breaks if we go above the limit.

wopu-ot commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -193,6 +193,23 @@ config BT_CTLR_TX_BUFFER_SIZE
+
+config BT_CTLR_ISO_TX_BUFFER_SIZE
+	int "Isochronous Tx buffer size"
+	range 27 251
```
I believe we need to include segmentation/fragmentation in the design from the beginning, to avoid making choices now that will hurt us later.

Thalley replied:
@wopu-ot That's probably a good idea :) 

I'm still a bit confused whether or not the maximum length can be `0xFFFF` or `0x7FFF`. What's your understanding of the maximum length?

wopu-ot replied:
I think you should ask Bluetooth-SIG for a clarification, not all corners of the specification are necessarily consistent ;-)

Thalley replied:
Mail sent to the core reflector, awaiting response.

Thalley replied:
@wopu-ot 
https://www.bluetooth.org/errata/errata_view.cfm?errata_id=16205

So apparently it's even odder than expected. The read size may return 0x0001-0xFFFF, the length field can only contain 0x0001-0x3FFF and the actual payload is seemingly limited to 0x0001-0x0FFF. 

For now it's set to maximum 0x3FFF. Should we reduce to it 0x0FFF, or await for the erratum?

wopu-ot replied:
I would reduce to 0x0FFF to avoid wasteful configurations. We can always adapt if the specification changes to make bigger buffer sizes useful.

Thalley replied:
Done

wopu-ot commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -28,9 +28,11 @@
 #include "hal/ccm.h"
 #include "ll_sw/pdu.h"
 #include "ll_sw/lll.h"
+#include "lll_adv.h"
```
We can look into this in another PR, but ideally the prefix should be there for all or for none.

Thalley replied:
The "issue" here is that there won't be a common prefix for this specific file I suppose, given that it's vendor specific.

wopu-ot replied:
I guess we could just drop the "ll_sw/" prefix. But again, I don't think this discussion needs to be part of this PR.

cvinayak commented:
please rebase.

Thalley committed:
Bluetooth: controller: Rearrange functions in ull_adv_iso.c

Rearranged the functions so the order of functions match
the common pattern in other ULL files.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: controller: Move ISO tx functionality to ull_iso.c

Moves the ISO tx pool and functions from ull_adv_iso.c to
ull_iso.c as the pool and functions will be shared between
BIS and CIS.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

wopu-ot reviewed (approved):
LGTM. #30287 should fix the failing CI, sorry for that.

Thalley committed:
Bluetooth: controller: ull_adv_iso merge unknown adv id check

Merges the 3 checks for checking valid adv when creating a big.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -10,28 +10,164 @@
+	struct node_tx *tx;
 
-	return BT_HCI_ERR_CMD_DISALLOWED;
+	adv_iso = ull_adv_iso_get(big_handle);
+
+	if (!adv_iso) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
```
If expecting `ull_adv_iso_is_created_get` to return NULL, then we'd have to call `ull_adv_iso_get` again to get a valid pointer. Instead I added a `adv-iso->is_created` check.

Thalley committed:
Bluetooth: controller: ll_big_create add check if handle already created

Adds a check to verify that the BIG handle isn't already allocated.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -193,6 +193,22 @@ config BT_CTLR_TX_BUFFER_SIZE
 	  Maximum is set to 251 due to implementation limitations (use of
 	  uint8_t for length field in PDU buffer structure).
 
+config BT_CTLR_ISO_TX_BUFFERS
+	int "Number of Isochronous Tx buffers"
+	default 3
```
Added depends on the ISO configs. Didn't add any in the first place, since the `BT_CTLR_TX_BUFFERS` doesn't depend on conn support either.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1377,22 +1379,37 @@ static void le_set_adv_enable(struct net_buf *buf, struct net_buf **evt)
 
 #if defined(CONFIG_BT_CTLR_ADV_ISO)
-static void le_create_big(struct net_buf *buf, struct net_buf **evt)
+static void le_create_big(struct net_buf *buf, struct net_buf **evt,
+			  void **node_rx)
```
Hmm, yeah you are right. I had imagined that the event could be returned after enqueing the ACAD and empty BIS PDU, but I think you are correct in the sense that we need to wait for the over-air transmit before returning the event.

Thalley committed:
Bluetooth: controller: removed node_rx for big create/terminate

Removed the node_rx argument for the BIG create/terminate
functions. The event should not be sent right after the
command, but needs to wait for transmissions of
empty BIS packets or BIG terminate indications.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: controller: Remove empty BIS packet tx from ULL

The empty BIS packet should be enqueue and handled in the LLL.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
Without knowing the innards of this structure and how it is used, I don't think that's something I can easily put together. I suggest that I add a TODO here, and then we can deal with optimization later.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -62,11 +198,235 @@ uint8_t ll_big_test_create(uint8_t big_handle, uint8_t adv_handle,
+
+	adv_iso->is_created = 0U;
+
+	/* Prepare BIG terminate event */
+	node_rx = (void *)&adv_iso->node_rx_terminate;
+	node_rx->hdr.type = NODE_RX_TYPE_BIG_TERMINATE;
+	node_rx->hdr.handle = big_handle;
+	node_rx->hdr.rx_ftr.param = adv_iso;
+	*((uint8_t *)node_rx->pdu) = reason;
+
+	*rx = node_rx;
+
```
Added a TODO for this in `tx_lll_flush`, as that needs to have some more LLL implemented as far as I can tell.

Thalley committed:
Bluetooth: controller: Guard ll_adv_iso_by_hci_handle_* functions

Guards the functions with CONFIG_BT_CTLR_HCI_ADV_HANDLE_MAPPING
and added a non-HCI inline function for them.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -62,11 +198,235 @@ uint8_t ll_big_test_create(uint8_t big_handle, uint8_t adv_handle,
+	return 0;
+}
+
+uint8_t ll_adv_iso_by_hci_handle_get(uint8_t hci_handle, uint8_t *handle)
```
Done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_types.h:
```diff
@@ -59,4 +59,41 @@ struct ll_adv_sync_set {
+
+	uint8_t  bcode[16];
+
+	struct node_rx_hdr node_rx_complete;
```
That doesn't seem necessary:
```
	node_rx->hdr.rx_ftr.param = adv_iso;
```

The only thing gained from the customs struct for terminate is to add the reason.

cvinayak replied:
Ah, true, rx_ftr is inside the header!

Thalley commented:
@cvinayak Fixed most of your comments. Some (non-trivial fixed) have been deferred to future work, as to not delay this PR too much, as some of them would require additional research for me, and I won't be able to allocate the time for this in the near future.

cvinayak reviewed (approved):
As further work is in progress, approving this PR and changes will be done in the future when LLL is developed for ISO.
